homeassistant/components/yeelight/__init__.py

```diff
@@ -139,73 +152,252 @@
+                entry,
+                data={
+                    CONF_DISCOVERY: entry.data.get(CONF_DISCOVERY),
+                    CONF_NAME: entry.data.get(CONF_NAME),
+                    CONF_IP_ADDRESS: entry.data[CONF_IP_ADDRESS],
+                },
+                options={
+                    CONF_MODEL: entry.data.get(CONF_MODEL, ""),
+                    CONF_TRANSITION: entry.data.get(
+                        CONF_TRANSITION, DEFAULT_TRANSITION
+                    ),
+                    CONF_MODE_MUSIC: entry.data.get(
+                        CONF_MODE_MUSIC, DEFAULT_MODE_MUSIC
+                    ),
+                    CONF_SAVE_ON_CHANGE: entry.data.get(
+                        CONF_SAVE_ON_CHANGE, DEFAULT_SAVE_ON_CHANGE
+                    ),
+                    CONF_NIGHTLIGHT_SWITCH: entry.data.get(
+                        CONF_NIGHTLIGHT_SWITCH, DEFAULT_NIGHTLIGHT_SWITCH
+                    ),
+                },
+            )
 
-    if DOMAIN in config:
-        for ipaddr, device_config in conf[CONF_DEVICES].items():
-            _LOGGER.debug("Adding configured %s", device_config[CONF_NAME])
-            _setup_device(hass, config, ipaddr, device_config)
+        config = {
+            CONF_NAME: entry.data.get(CONF_NAME),  # Support name import from yaml
+            **entry.options,
+        }
+        await _async_setup_device(hass, entry.data[CONF_IP_ADDRESS], config)
+        unset = entry.add_update_listener(_async_update_listener)
+        hass.data[DOMAIN][DATA_CONFIG_ENTRIES][entry.entry_id] = {
+            DATA_UNSUB_UPDATE_LISTENER: unset,
+        }
+        for component in PLATFORMS:
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(entry, component)
+            )
 
     return True
 
 
-def _setup_device(hass, _, ipaddr, device_config):
-    devices = hass.data[DATA_YEELIGHT]
-
-    if ipaddr in devices:
-        return
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok:
+        if entry.data[CONF_DISCOVERY]:
+            data = hass.data[DOMAIN][DATA_CONFIG_ENTRIES].pop(entry.entry_id)
+            scanner = data[DATA_SCANNER]
+            await scanner.stop_scan()
+            for ipaddr in scanner.seen:
+                hass.data[DOMAIN][DATA_DEVICES].pop(ipaddr)
+            data[DATA_UNSUB_UPDATE_LISTENER]()
+        else:
+            hass.data[DOMAIN][DATA_DEVICES].pop(entry.data[CONF_IP_ADDRESS])
+            data = hass.data[DOMAIN][DATA_CONFIG_ENTRIES].pop(entry.entry_id)
+            data[DATA_UNSUB_UPDATE_LISTENER]()
+
+    return unload_ok
+
+
+async def _async_setup_device(
+    hass: HomeAssistant,
+    ipaddr: str,
```

Comment:
Please separate words with underscore in variable names.
```suggestion
    ip_addr: str,
```

Reply:
`ipaddr` is used all over the place in this integration. I'll make this change and some other code improvements in a followup PR.